Lib/concurrent/futures/_base.py

```diff
@@ -170,6 +170,17 @@ def _create_and_install_waiters(fs, return_when):
 
     return waiter
 
+
+def _yield_future(fs, waiter, ref_collect=()):
+    while fs:
+        with fs[0]._condition:
```

Comment:
If your question is about line 176.  Based on  issue20319, changes on future waiters list should be locked.